### PR TITLE
Wait for confirms

### PR DESCRIPTION
**Pull request by carlhoerberg**
Always wait for all outstanding confirms.

### CODE REVIEW

File: shard.yml [hunk 1]
```diff
@@ -7,6 +7,7 @@ authors:
 
 dependencies:
   amq-protocol:
+    #path: ../amq-protocol.cr
```
Comment (spuun) [inline shard.yml:10]:
```suggestion
```
A tip here is to use `shard.override.yml` (which should be added to .gitignore)

Comment (carlhoerberg) [reply shard.yml:10]: thanks! didn't know of it

File: src/amqp-client/channel.cr [hunk 2]
```diff
@@ -255,25 +254,44 @@ class AMQP::Client
...
-      msgid = basic_publish(io, bytesize, exchange, routing_key, mandatory, immediate, properties)
-      wait_for_confirm(msgid)
+      basic_publish(io, bytesize, exchange, routing_key, mandatory, immediate, properties)
+      wait_for_confirms
+    end
+
+    # Returns when there are no unconfirmed publishes on the channel
+    # Raises if there was any negative acknowledgements
+    def wait_for_confirms : Bool
+      ok = @unconfirmed_empty.receive
```
Comment (carlhoerberg) [inline src/amqp-client/channel.cr:282]: wonder if we should check if the dequeue is empty first? could get stuck here if the user hasn't called this method while the ack comes back from the server, say: 

```crystal
ch.confirm_select
ch.basic_publish "hello world"
sleep 1
ch.wait_for_confirms # will hang forever
```

or just
```crystal
ch = connection.channel
ch.wait_for_confirms # will hang forever
```

File: src/amqp-client/channel.cr [hunk 3]
```diff
@@ -286,32 +304,51 @@ class AMQP::Client
...
+          end
+        end
+        @last_confirm = {delivery_tag, acked}
+        if @unconfirmed_publishes.empty?
+          loop do
+            select
+            when @unconfirmed_empty.send acked
+            else
+              break
+            end
```
Comment (carlhoerberg) [inline src/amqp-client/channel.cr:349]: how could `unconfirm_publishes` be empty here? that should be an error/exception, no?

File: src/amqp-client/channel.cr [hunk 4]
```diff
@@ -286,32 +304,51 @@ class AMQP::Client
...
+    @unconfirmed_empty = ::Channel(Bool).new
+    @confirm_lock = Mutex.new
     @last_confirm = {0_u64, true}
 
+    @[Deprecated("Use `#wait_for_confirms` instead")]
     def on_confirm(msgid, &blk : Bool -> Nil)
       raise ArgumentError.new "Confirm id must be > 0" unless msgid > 0
       last_confirm, last_confirm_ok = @last_confirm
       if last_confirm >= msgid.to_u64
         blk.call last_confirm_ok
```
Comment (carlhoerberg) [inline src/amqp-client/channel.cr:317]: this assumes that ack's are coming in order, so it's probably a fault implementation. a first step could be to check if the last `ack` was sent with the `multiple` flag, then this holds.

File: src/amqp-client/channel.cr (hunk 3 above)
Comment (baelter) [reply src/amqp-client/channel.cr:349]: We just emptied it on the line just above?

File: src/amqp-client/channel.cr (hunk 4 above)
Comment (baelter) [reply src/amqp-client/channel.cr:317]: Right, we have this bug in the current release. I'll always wait for all confirms to keep it simple.

**Pull request closed without merging**